wp-components-php/inc/classes/class-render-controller.php

```diff
@@ -0,0 +1,200 @@
+<?php
+/**
+ * Render_Controller class file.
+ *
+ * @package WP_Component
+ */
+
+namespace WP_Component\PHP;
+
+/**
+ * Render Controller.
+ */
+class Render_Controller {
+
+	/**
+	 * Set the default cache TTL to 15 minutes for cached partials.
+	 *
+	 * @var integer
+	 */
+	public $default_cache_ttl = 900;
+
+	/**
+	 * Holds references to the singleton instances.
+	 *
+	 * @var array
+	 */
+	private static $instance;
+
+	/**
+	 * Unused.
+	 */
+	private function __construct() {
+		// Don't do anything, needs to be initialized via instance() method.
+	}
+
+	/**
+	 * Get an instance of the class.
+	 *
+	 * @return Render_Controller
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new static();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * The template stack.
+	 *
+	 * @var array
+	 */
+	public $stack = array();
+
+	/**
+	 * The currently-active renderable.
+	 *
+	 * @var Renderable
+	 */
+	public $current_renderable;
+
+	/**
+	 * Render a component and its children
+	 *
+	 * @param \Wp_Component\Component $component_instance Instance of a component (or template) to render
+	 * @param bool                    $return             Should this component's markup be return instead of
+	 *                                                    printed?
+	 * @param bool|array              $cache              If set, the template part will be cached. If
+	 *                                                    true, results will be cached for
+	 *                                                    {@see Render_Controller::$default_cache_ttl} and
+	 *                                                    the transient will be generated from this
+	 *                                                    variable. Optionally, either can be set by
+	 *                                                    passing an array with 'key' and/or 'ttl' keys.
+	 */
+	public function render( $component_instance, $return = false, $cache = false ) {
+		if ( $cache ) {
+
+			if ( is_bool( $cache ) ) {
+				$cache = array();
+			}
+
+			// If no key was provided, make one.
+			if ( empty( $cache['key'] ) ) {
+				$cache['key'] = self::cache_key( $args );
+			}
+
+			// If no TTL was supplied, set a default.
+			if ( ! isset( $cache['ttl'] ) ) {
+				$cache['ttl'] = $this->default_cache_ttl;
+			}
+
+			// If we have a cache hit, serve it.
+			$renderable = get_transient( $cache['key'] );
+			if ( false !== $renderable ) {
+				if ( $return ) {
+					return $renderable;
+				} else {
+					echo $renderable; // wpcs: xss ok.
```

Comment:
This style of ignoring a run is considered deprecated, we should use `// phpcs: ignore NameOfRule`